lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/AbstractTestS3FileSystem.java

```diff
@@ -167,29 +172,92 @@ void testFileWithTrailingWhitespaceAgainstNativeClient()
                 fileSystem.deleteDirectory(data);
-                assertThat(fileSystem.listDirectories(getRootLocation())).containsExactly(data);
+                assertThat(fileSystem.listDirectories(getRootLocation())).isEmpty();
 
                 fileSystem.deleteDirectory(getRootLocation());
-                assertThat(fileSystem.listDirectories(getRootLocation())).containsExactly(data);
+                assertThat(fileSystem.listDirectories(getRootLocation())).isEmpty();
             }
             finally {
                 s3Client.deleteObject(delete -> delete.bucket(bucket()).key(key));
             }
         }
     }
+
+    @Test
+    void testDeleteEmptyDirectoryWithDeepHierarchy()
+            throws IOException
+    {
+        try (S3Client s3Client = createS3Client(); Closer closer = Closer.create()) {
+            createDirectory(closer, s3Client, "deep/dir");
+            createBlob(closer, "deep/dir/file1.txt");
+            createBlob(closer, "deep/dir/file2.txt");
+            createBlob(closer, "deep/dir/file3.txt");
+            createDirectory(closer, s3Client, "deep/dir/dir4");
+            createBlob(closer, "deep/dir/dir4/file5.txt");
+
+            assertThat(fileSystem.listFiles(getRootLocation()).hasNext()).isTrue();
+
+            Location directory = getRootLocation().appendPath("deep/dir/");
+            assertThat(fileSystem.listDirectories(getRootLocation().appendPath("deep"))).containsExactly(directory);
+            assertThat(fileSystem.listDirectories(directory)).containsExactly(getRootLocation().appendPath("deep/dir/dir4/"));
+
+            fileSystem.deleteDirectory(directory);
+            assertThat(fileSystem.listDirectories(getRootLocation().appendPath("deep"))).isEmpty();
+            assertThat(fileSystem.listDirectories(getRootLocation())).isEmpty();
+            assertThat(fileSystem.listFiles(getRootLocation()).hasNext()).isFalse();
+        }
+    }
+
+    protected Location createDirectory(Closer closer, S3Client s3Client, String path)
+    {
+        Location location = createLocation(path);
+        closer.register(new TempDirectory(s3Client, path)).create();
+        return location;
+    }
+
+    protected class TempDirectory
+            implements Closeable
+    {
+        private final S3Client s3Client;
+        private final String path;
+
+        public TempDirectory(S3Client s3Client, String path)
+        {
+            this.s3Client = requireNonNull(s3Client, "s3Client is null");
+            String key = requireNonNull(path, "path is null");
+            this.path = key.endsWith("/") ? key : key + "/";
+        }
+
+        public void create()
+        {
+            PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                    .bucket(bucket())
+                    .key(path)
+                    .contentLength(0L)
+                    .metadata(Map.of("ContentType", "application/x-directory"))
```

Comment:
We shouldn't need to set this. The code should work with any file that ends in slash. The original code that creates a zero length file should be sufficient.